src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -13690,6 +13690,9 @@ void Unit::UpdateCharmAI()
             delete i_AI;
             i_AI = i_disabledAI;
             i_disabledAI = NULL;
+
+            if (GetTypeId() == TYPEID_UNIT)
```

Comment:
Oh, now I get it. Right when the AI change is restored, u call the oncharm hook. Smart.